File: MultiboxClientCSharpCore/Model/ServerPlaylistShadow.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using CSharpNetInterface.Util;
using MultiboxClientCSharpCore.Net;
using MultiboxClientCSharpCore.Net.ServerInterfaceEventArgs;
using MultiboxClientCSharpCore.Model.PlaylistEventArgs;

namespace MultiboxClientCSharpCore.Model
{
	class ServerPlaylistShadow : IPlaylist
	{
		private readonly IServerInterface serverInterface;
		private readonly IMessageQueue messageQueue;

		private IList<MultimediaItem> actualPlaylist;

		public event EventHandler<PlaylistChangedEventArgs> PlaylistChanged = delegate { };
		public event EventHandler<ItemAddedEventArgs> ItemAdded = delegate { };

		private PeriodicWorkDispatcher stateChecker;

		public ServerPlaylistShadow(IServerInterface serverInterface, IMessageQueue messageQueue)
		{
			this.serverInterface = serverInterface;
			this.messageQueue = messageQueue;

			actualPlaylist = new List<MultimediaItem>();

			this.stateChecker = new PeriodicWorkDispatcher(messageQueue, 5000, UpdatePlaylist);
		}

		public void Init()
		{
			serverInterface.PlaylistReceived += OnPlaylistReceived;
			serverInterface.AddingLibraryItemToPlaylistResult += OnAddingLibraryItemToPlaylistResult;

			UpdatePlaylist();

			stateChecker.Start();
		}

		public void Close()
		{
			serverInterface.PlaylistReceived -= OnPlaylistReceived;
			serverInterface.AddingLibraryItemToPlaylistResult -= OnAddingLibraryItemToPlaylistResult;

			stateChecker.Stop();
		}

		private void UpdatePlaylist()
		{
			serverInterface.RequestPlaylistUpdate();
		}

		public IList<MultimediaItem> Items
		{
			get { return new List<MultimediaItem>(actualPlaylist); }
		}

		public void AddItem(long itemId)
		{
			serverInterface.AddLibraryItemToPlaylist(itemId);
		}

		private void OnPlaylistReceived(object sender, PlaylistReceivedEventArgs eventArgs)
		{
			UpdatePlaylist(eventArgs.Playlist);
		}

		private void UpdatePlaylist(IList<MultimediaItem> playlist)
		{
			actualPlaylist.Clear();
			foreach (MultimediaItem multimedia in playlist)
			{
				actualPlaylist.Add(multimedia);
			}

			PlaylistChangedEventArgs eventArgs = new PlaylistChangedEventArgs(new List<MultimediaItem>(actualPlaylist));
			messageQueue.Post(() => PlaylistChanged(this, eventArgs));
		}

		private void OnAddingLibraryItemToPlaylistResult(object sender, AddingLibraryItemToPlaylistResultEventArgs eventArgs)
		{
			bool success = eventArgs.Result;
			MultimediaItem multimedia = eventArgs.Multimedia;

			if (actualPlaylist.Count <= 1)
			{
				return;
			}

			IList<MultimediaItem> newPlaylist = new List<MultimediaItem>(actualPlaylist);
			newPlaylist.Add(multimedia);

			UpdatePlaylist(newPlaylist);

			ItemAddedEventArgs itemAddedEventArgs = new ItemAddedEventArgs(success, multimedia);
			messageQueue.Post(() => ItemAdded(this, itemAddedEventArgs));
		}
	}
}

File: MultiBoxClientJavaCore/src/sk/hackcraft/multibox/model/ServerPlaylistShadow.java
package sk.hackcraft.multibox.model;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

import sk.hackcraft.multibox.model.libraryitems.MultimediaItem;
import sk.hackcraft.multibox.net.ServerInterface;
import sk.hackcraft.multibox.net.ServerInterface.ServerInterfaceEventAdapter;
import sk.hackcraft.util.MessageQueue;

public class ServerPlaylistShadow implements Playlist
{
	private final ServerInterface serverInterface;
	private final MessageQueue messageQueue;
	
	private List<MultimediaItem> actualPlaylist;
	
	private List<Playlist.PlaylistEventListener> playlistListeners;
	
	private ServerListener serverListener;
	
	public ServerPlaylistShadow(ServerInterface serverInterface, MessageQueue messageQueue)
	{
		this.serverInterface = serverInterface;
		this.messageQueue = messageQueue;
		
		playlistListeners = new LinkedList<Playlist.PlaylistEventListener>();
		
		serverListener = new ServerListener();
		
		actualPlaylist = new LinkedList<MultimediaItem>();
	}
	
	@Override
	public void init()
	{
		serverInterface.registerEventListener(serverListener);
		serverInterface.requestPlaylistUpdate();
	}

	@Override
	public void close()
	{
		serverInterface.unregisterEventListener(serverListener);
	}
	
	@Override
	public void registerListener(PlaylistEventListener listener)
	{
		playlistListeners.add(listener);
	}
	
	@Override
	public void unregisterListener(PlaylistEventListener listener)
	{
		playlistListeners.remove(listener);
	}

	@Override
	public List<MultimediaItem> getItems()
	{
		return Collections.unmodifiableList(actualPlaylist);
	}
	
	@Override
	public void addItem(long itemId)
	{
		serverInterface.addLibraryItemToPlaylist(itemId);
	}
	
	private void updatePlaylist(List<MultimediaItem> playlist)
	{
		actualPlaylist.clear();
		actualPlaylist.addAll(playlist);
		
		final List<MultimediaItem> playlistCopy = new LinkedList<MultimediaItem>(actualPlaylist);
		
		for (final Playlist.PlaylistEventListener listener : playlistListeners)
		{
			messageQueue.post(new Runnable()
			{
				@Override
				public void run()
				{
					listener.onPlaylistChanged(playlistCopy);
				}
			});
		}
	}
	
	private class ServerListener extends ServerInterfaceEventAdapter
	{
		@Override
		public void onPlaylistReceived(List<MultimediaItem> playlist)
		{
			updatePlaylist(playlist);
		}
		
		@Override
		public void onAddingLibraryItemToPlaylistResult(final boolean result, final MultimediaItem multimedia)
		{
			List<MultimediaItem> updatedPlaylist = new LinkedList<MultimediaItem>(actualPlaylist);
			updatedPlaylist.add(multimedia);
			
			updatePlaylist(updatedPlaylist);
			
			for (final Playlist.PlaylistEventListener listener : playlistListeners)
			{
				messageQueue.post(new Runnable()
				{
					@Override
					public void run()
					{
						listener.onItemAdded(result, multimedia);
					}
				});
			}
		}
	}
}
